### ISSUE TITLE
The code of OFA-base is inconsistent with the pre-trained checkpoint

### ISSUE THREAD

**Issue by zzhanghub** [OFA-Sys/OFA#229]
Thanks for your awesome work. 
Something has bothered me recently. When I continued to train OFA-base (I tried to collect all the pre-training data of OFA), I found that a few training steps (10 steps) would make the performance of OFA-base worse. 
I checked the config in [checkpoint](https://github.com/OFA-Sys/OFA/blob/main/checkpoints.md) and the config in [pretrain_ofa_base.sh](https://github.com/OFA-Sys/OFA/blob/main/run_scripts/pretraining/pretrain_ofa_base.sh), and found many differences. What might affect the results?

In addition, I found that there is a dimension inconsistency in the network.  `decoder.image_position_idx": "<class 'torch.Tensor'> torch.Size([1026])` in code and `decoder.image_position_idx": "<class 'torch.Tensor'> torch.Size([1025])` in ckpt.
Is this the reason for the decline of performance?

**Comment by zzhanghub** [comment]
I found that if I load the pre-trained checkpoint, save it, and then load it (there is no training during this period). The results will be very different. I don't know what parameter caused this problem.

**Comment by Charles-Xie** [comment]
I also came across such problem (the zero-shot performance of the official checkpoint drops if we load and save it (without training)).
I try to compare the hyper-parameters in the official checkpoint with those in our checkpoint (saved after loading the official checkpoint). We found the shape of the image_position_idx to be different (the original: 1025, the loaded-and-saved: 1026). Is this the possible reason for the perform drop?

**Comment by zzhanghub** [comment]
> I also came across such problem (the zero-shot performance of the official checkpoint drops if we load and save it (without training)). I try to compare the hyper-parameters in the official checkpoint with those in our checkpoint (saved after loading the official checkpoint). We found the shape of the image_position_idx to be different (the original: 1025, the loaded-and-saved: 1026). Is this the possible reason for the perform drop?

I found a commit： https://github.com/OFA-Sys/OFA/commit/f59bc9492eca2bc5f883dc78485298b1af149c8c
This affects the dimension. However, I change the `769` to `768`. The results did not change.

Do you have any other ideas?

**Comment by zzhanghub** [comment]
Some detailed differents between resaved ckpt and [ofa-base ckpt](https://github.com/OFA-Sys/OFA/blob/main/checkpoints.md) are below.
1. Our training task removes MLM and images infilling. (From the paper tab.10, it should not be the main reason. Moreover, we only trained 10 steps, and the model only sees dozens of samples);
2. Sample in training script_ patch_ Num = 196, but 224 in [ofa-base ckpt](https://github.com/OFA-Sys/OFA/blob/main/checkpoints.md). (we set it to be 224, but it still has serious performance degradation after 10 steps);
3. The current code and cpkt are inconsistent in some model dimensions. In the latest model of repo, the `decoder image_ position_ idx": <class 'torch. Tensor'> torch. The size ([1026])` is 1026, but the parameter in the CKPT of ofa-base is 1025. The [commit](https://github.com/OFA-Sys/OFA/commit/f59bc9492eca2bc5f883dc78485298b1af149c8c) is the cause of inconsistency; (we **manually** set it to be consistent, but it still has serious performance degradation after 10 steps);
5. [pretrain_ofa_base.sh](https://github.com/OFA-Sys/OFA/blob/main/run_scripts/pretraining/pretrain_ofa_base.sh) includes `neg_sample_dir`, while `neg_sample_dir` is null in  [ofa-base ckpt](https://github.com/OFA-Sys/OFA/blob/main/checkpoints.md) .
6. It seems that the learning rate is different, but we tried a variety of LR, and the results were similar.

We look forward to your reply!🙏  Thank you!!!

**Comment by logicwong** [comment]
@zzhanghub @Charles-Xie Changing the dimension of `image_position_idx` to 1026 is to be consistent with the dimension of `embed_positions`, it shouldn't affect the performance. I guess you save the model during training, but even if the model parameters are not updated, the parameters of `batch_norm` will still be updated and degenerate the performance. You can set `--freeze-resnet` to fix the problem.

@zzhanghub For detailed differents between resaved ckpt and ofa-base ckpt:
1. That's ok.
2. In order to improve the pre-training, we use a larger sample_patch_num in the later stage of training. But in recent experiments, we found that using --patch-image-size=256 without sampling patches is better.
3. As mentioned above, this does not affect performance。
4. The path of `neg_sample_dir` has some sensitive information, so we manually deleted it before releasing ckpt.
5. The model was interrupted many times during pre-training, we reset the learning rate a few times during this period,  but this should have little effect on the performance

**Comment by zzhanghub** [comment]
> @zzhanghub @Charles-Xie Changing the dimension of `image_position_idx` to 1026 is to be consistent with the dimension of `embed_positions`, it shouldn't affect the performance. I guess you save the model during training, but even if the model parameters are not updated, the parameters of `batch_norm` will still be updated and degenerate the performance. You can set `--freeze-resnet` to fix the problem.
> 
> @zzhanghub For detailed differents between resaved ckpt and ofa-base ckpt:
> 
> 1. That's ok.
> 2. In order to improve the pre-training, we use a larger sample_patch_num in the later stage of training. But in recent experiments, we found that using --patch-image-size=256 without sampling patches is better.
> 3. As mentioned above, this does not affect performance。
> 4. The path of `neg_sample_dir` has some sensitive information, so we manually deleted it before releasing ckpt.
> 5. The model was interrupted many times during pre-training, we reset the learning rate a few times during this period,  but this should have little effect on the performance

Thanks for your patient reply. We try the `--freeze-resnet` option yesterday. We found that the decline did relieve when reloading saved ckpt. However, after some steps training, the performance will still decline significantly (The performance of the model is reduced to one-third of the original). 


1. Are there any other key parameters or settings that will affect the pre-trained performance?
2. We use the `neg_sample_dir` provided in the [[dataset page.](https://github.com/OFA-Sys/OFA/blob/main/datasets.md)](https://github.com/OFA-Sys/OFA/blob/main/datasets.md) Is this folder all the negative sample data you use? Is this the cause of the performance degradation?

**Comment by zzhanghub** [comment]
> @zzhanghub @Charles-Xie Changing the dimension of `image_position_idx` to 1026 is to be consistent with the dimension of `embed_positions`, it shouldn't affect the performance. I guess you save the model during training, but even if the model parameters are not updated, the parameters of `batch_norm` will still be updated and degenerate the performance. You can set `--freeze-resnet` to fix the problem.
> 
> @zzhanghub For detailed differents between resaved ckpt and ofa-base ckpt:
> 
> 1. That's ok.
> 2. In order to improve the pre-training, we use a larger sample_patch_num in the later stage of training. But in recent experiments, we found that using --patch-image-size=256 without sampling patches is better.
> 3. As mentioned above, this does not affect performance。
> 4. The path of `neg_sample_dir` has some sensitive information, so we manually deleted it before releasing ckpt.
> 5. The model was interrupted many times during pre-training, we reset the learning rate a few times during this period,  but this should have little effect on the performance

Thanks for your patient reply. We try the `--freeze-resnet` option yesterday. We found that the decline did relieve when reloading saved ckpt. However, after some steps of training, the performance will still decline significantly (The performance of the model is reduced to one-third of the original). 


1. Are there any other key parameters or settings that will affect the pre-trained performance?
2. We use the `neg_sample_dir` provided in the [[dataset page.](https://github.com/OFA-Sys/OFA/blob/main/datasets.md)](https://github.com/OFA-Sys/OFA/blob/main/datasets.md) Is this folder all the negative sample data you use? Is this the cause of the performance degradation?

**Comment by logicwong** [comment]
@zzhanghub There could be many reasons:
1. We use Pillow to open and save the picture, and the picture information will be lost in this process. I don't know if you take the same way to process the pictures, if not, the data we use will be different from yours;
2. Since ofa-base has converged, if you continue pretrain it with a large learning rate, the model may jump out of the optimal point, so you will observe the loss rises;
3. Batch size matters, since we don't use sync bn, the per-gpu-batch-size will affect the batch norm, thereby affecting the stability of training. You can use `--freeze-resnet` to stabilize the training, but this will affect the performance, especially in case 1.

So I suggest:
1. Use large per-gpu-batch-size or use sync bn;
2. Tune the hyperparameters, and choose an appropriate learning rate to restart training for a longer time. In the beginning you may observe loss rises, but then the loss should decrease and converge to a stable value.

**Comment by zzhanghub** [comment]
> @zzhanghub There could be many reasons:
> 
> 1. We use Pillow to open and save the picture, and the picture information will be lost in this process. I don't know if you take the same way to process the pictures, if not, the data we use will be different from yours;
> 2. Since ofa-base has converged, if you continue pretrain it with a large learning rate, the model may jump out of the optimal point, so you will observe the loss rises;
> 3. Batch size matters, since we don't use sync bn, the per-gpu-batch-size will affect the batch norm, thereby affecting the stability of training. You can use `--freeze-resnet` to stabilize the training, but this will affect the performance, especially in case 1.
> 
> So I suggest:
> 
> 1. Use large per-gpu-batch-size or use sync bn;
> 2. Tune the hyperparameters, and choose an appropriate learning rate to restart training for a longer time. In the beginning you may observe loss rises, but then the loss should decrease and converge to a stable value.

@logicwong 
Thank you for your reply. Recently, we dived into the codes and found some details that confused me:
1. Why use `T.ObjectCenterCrop` for VG tasks (see [codes](https://github.com/OFA-Sys/OFA/blob/1809b55687458dbb4b1686a89b569f0efb2536e3/data/pretrain_data/unify_dataset.py#L231))? This operation makes target object in the center of the picture, reducing the difficulty of learning, but, in my view, it will affect the grounding ability of zero-shot.
2. Why use `padding` for CC12M (see [codes](https://github.com/OFA-Sys/OFA/blob/1809b55687458dbb4b1686a89b569f0efb2536e3/data/pretrain_data/unify_dataset.py#L321))? 

We look forward to your reply! Thanks again!

**Comment by logicwong** [comment]
@zzhanghub 

1. If we simply use `T.CenterCrop` after `T.RandomResize`, the processed image may not contain the target object, so we use `T.ObjectCenterCrop`. Another way is to directly resize the image to `patch_image_size` without any cropping, you can also try this.
2. CC12M is noisier than other caption datasets such as COCO and CC3M, the first few words of each caption in CC12M may be unpredictable, so we use padding to remind the model not to calculate the loss for the first two words. This trick is somewhat similar to PrefixLM.

**Comment by zzhanghub** [comment]
Thank you !!!
